Add log following

timbrown5 opened the pull request:
Extension of: #175 

Video:
https://user-images.githubusercontent.com/4447764/225406899-deae7b82-b1e2-4bcb-bc8e-f3717db6772f.mov


Changes:
- Add a new ConsoleLogStream component to house the Virtuoso list and logic.
-- Allow the Virtuoso list to follow logs and request more output (when following).
--- This changes the behaviour a bit so that the Plugin will only request new log output when following the log. If we add a search feature this might need a re-think but it simplifies things.
- Updated onSelect code to expand a newly focused stage and expand the last step card.
- The Console will stick on the selected node. The node can be unselected by clicking it again, at which point the console will follow the log again.

I've been using this Pipeline for testing:
```
pipeline {
    agent any

    stages {
        stage('Hello') {
            steps {
                script {
                    for(int i=1; i<=10; i++) {
                        stage("Stage - ${i}") {
                            sh 'for i in {1..120}; do sleep 1; echo "Slept ${i} times"; done'
                        }
                    }
                }
            }
        }
    }
}
```

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira

timja commented on src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx:
```diff
@@ -458,4 +472,38 @@ export default class PipelineConsole extends React.Component<
+    }
+  }
+
+  zzrender() {
```
is this intentional?

timja reviewed (commented):
Can you add some padding / margin to the scroll to bottom text please?
----

<img width="156" alt="image" src="https://user-images.githubusercontent.com/21194782/225437782-1b4daa4b-29ba-4136-a696-61d32aef63e0.png">

----

Unrelated but noticed while testing, as you get to 3 digit line numbers there's no enough padding it's touching the side:

<img width="424" alt="image" src="https://user-images.githubusercontent.com/21194782/225437846-182e7e70-de01-4029-bc6f-b4791dc0e2f7.png">

----

Is it worth adding some sort of spinner or loading bar to show that the step is still running / loading text? Currently the only way you know it's still doing something is when new lines appear

timbrown5 commented:
Sorry for the slow reply. I am adding some tests to the Pipeline view,
which required a small refactor to make parts easily mockable.
This has broken the polling so will put up an update once that's fixed.
One the plus side I think I've resolved your comments on my local copy.

On Wed, 15 Mar 2023, 20:49 Tim Jacomb, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> Can you add some padding / margin to the scroll to bottom text please?
>
> [image: image]
> <https://user-images.githubusercontent.com/21194782/225437782-1b4daa4b-29ba-4136-a696-61d32aef63e0.png>
> ------------------------------
>
> Unrelated but noticed while testing, as you get to 3 digit line numbers
> there's no enough padding it's touching the side:
>
> [image: image]
> <https://user-images.githubusercontent.com/21194782/225437846-182e7e70-de01-4029-bc6f-b4791dc0e2f7.png>
> ------------------------------
>
> Is it worth adding some sort of spinner or loading bar to show that the
> step is still running / loading text? Currently the only way you know it's
> still doing something is when new lines appear
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/253#pullrequestreview-1342403696>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABB54FBOTYWCILQDXRZZW3TW4ITL5ANCNFSM6AAAAAAV4G5BQM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>

timja reviewed (commented):
3 issues I noticed.

1. certain times overflow and touch the expander, a lot more likely to be noticed while watching the time increase:
<img width="134" alt="image" src="https://user-images.githubusercontent.com/21194782/226550591-08e26356-aec5-4905-8b87-b85728e818f5.png">
2. it's not possible to look at other logs while the job is running, it force jumps you to watch the running logs
3. when it force jumps you it doesn't update the active selection on the left panel (see stage 4 active but stage 8 logs)
<img width="1184" alt="image" src="https://user-images.githubusercontent.com/21194782/226550950-e8441bc2-c987-4bea-b478-ec6906b579f2.png">

timbrown5 commented:
> 3 issues I noticed.
> 
>     1. certain times overflow and touch the expander, a lot more likely to be noticed while watching the time increase:
> 
> 
> <img alt="image" width="134" src="https://user-images.githubusercontent.com/21194782/226550591-08e26356-aec5-4905-8b87-b85728e818f5.png">
> 2. it's not possible to look at other logs while the job is running, it force jumps you to watch the running logs 3. when it force jumps you it doesn't update the active selection on the left panel (see stage 4 active but stage 8 logs)
> 
> <img alt="image" width="1184" src="https://user-images.githubusercontent.com/21194782/226550950-e8441bc2-c987-4bea-b478-ec6906b579f2.png">

I'll have a look. The focus/selected node logic might be a little broken.
I've added the idea of a focused stage - the one open vs a selected stage - the one selected by the user. If the code is changing the selected stage that is probably an issue.

timbrown5 commented:
> 2. it's not possible to look at other logs while the job is running, it force jumps you to watch the running logs.
I think this hasn't been resolved (I fixed some issues with the step selection - and added tests - but not the stage selection).

> 3. when it force jumps you it doesn't update the active selection on the left panel (see stage 4 active but stage 8 logs).
I think the solution to this might resolve the behaviour @stuartrowe reported here:
https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/177#issuecomment-1297575955
Seems the workaround is to manually find the DOM element and focus it yourself. My only worry is that this could cause some unwanted behaviour - changing the user's focus when they are looking at something else (although as there are no input boxes this is likely not a big issue?).

I'm considering update the new behaviour to:
1. Select the TreeItem onNodefocus - this should remove the difference between focused and selected items (the focused item will always be selected).
1. Remove the 'focus' highlight - and move it to 'selected'. This means the highlight will remain even if the user clicks outside of the TreeView.
- Together, these should avoid the limitation @stuartrowe reported here (and 3 above):
https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/177#issuecomment-1297575955
3. If a TreeItem is selected then stick to that item.
1. Allow the user to click the selected item again to unselect it (the highlight will be removed). This will start following the Pipeline again on the next update - won't have an effect when the Pipeline is finished.

Depending on what unselecting an item does, we might need to add a custom CSS class to control highlighting.
Thoughts welcome :).

timja commented:
Sounds reasonable,

> Allow the user to click the selected item again to unselect it (the highlight will be removed). This will start following the Pipeline again on the next update - won't have an effect when the Pipeline is finished.

❤️

timbrown5 commented:
One issue is that `onSelected` is only fired when the selected field is set (not when it's cleared). Might need to add a custom clickable action to implement the above.

timbrown5 commented:
> One issue is that `onSelected` is only fired when the selected field is set (not when it's cleared). Might need to add a custom clickable action to implement the above.

This statement is incorrect. Don't know what was happening to make it only work 1/2 the time. It's working for me now.

timja commented:
Is this ready for re-review?

timbrown5 commented:
Sorry, had another look and found a couple of things that needed fixing (including your first issue above).

timja commented on src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx:
```diff
@@ -296,46 +398,30 @@ export default class PipelineConsole extends React.Component<
   }
 
   handleStepToggle(event: React.SyntheticEvent<{}>, nodeId: string): void {
+    console.log(`Node '${nodeId}' toggled.`);
```
remove?
```suggestion
```

timja reviewed (approved):
Only thing I noticed when testing is times can still quite easily get cut-off.
<img width="684" alt="image" src="https://user-images.githubusercontent.com/21194782/230319380-0b74bcfb-6fb9-4780-8944-baaa1682d701.png">

That's a Mac 13" at full width so a very standard size.

Nonblocking and can be improved later.

----

LGTM
